### PR TITLE
Rebase #177 - Shared breakpoints support

### PR DESCRIPTION
**Pull request by myihsan**
# Motivation and Context
XcodeGen does not support shared breakpoints.
There is a pull request #177 to support shared breakpoints but never update since 2017.

# Description
Rebase #177, and make some changes according to comments.

I am not sure if this is a proper way to contribute base on a long open pull request. If not, please tell me an appropriate way.

### CODE REVIEW

File: Docs/ProjectSpec.md
Comment (yonaskolb) [inline Docs/ProjectSpec.md]:
```suggestion
    - `FileBreakpoint`: file breakpoint
```

File: Tests/XcodeGenKitTests/SpecLoadingTests.swift
Comment (yonaskolb) [inline Tests/XcodeGenKitTests/SpecLoadingTests.swift]: Can we add test decoding all the supported properties in another section

File: Docs/ProjectSpec.md
Comment (yonaskolb) [inline Docs/ProjectSpec.md]: While all of these properties are currently strings in `XcodeProj` we should handle this better and make the relevant ones ints like:
- ignoreCount
- startingColumn
- endingColumn
- startingLine
- endingLine

The BreakpointGenerator can then turn these into strings for now

File: Docs/ProjectSpec.md
Comment (yonaskolb) [inline Docs/ProjectSpec.md]: You can see this `ignoreCount` has been set to an int which makes sense, hence the above comment

File: Sources/ProjectSpec/Breakpoint.swift
Comment (yonaskolb) [inline Sources/ProjectSpec/Breakpoint.swift]: Let's change relevant properties to `Int` as per comment on ProjectSpec, both for yaml/json parsing and also for easier Swift usage

File: Docs/ProjectSpec.md
Comment (yonaskolb) [inline Docs/ProjectSpec.md]: Do we need locations if they are not used?

File: Sources/ProjectSpec/Breakpoint.swift
Comment (yonaskolb) [inline Sources/ProjectSpec/Breakpoint.swift]: Maybe for now we can create 2 typealiases that we can use throughout the codebase
```
typealias BreakpointActionType XCBreakpointList.BreakpointProxy.BreakpointContent.BreakpointActionProxy.ActionExtensionID
typealias BreakpointType = XCBreakpointList.BreakpointProxy.BreakpointExtensionID
```

File: Sources/ProjectSpec/Breakpoint.swift
Comment (yonaskolb) [inline Sources/ProjectSpec/Breakpoint.swift]: Again, do we need this type?

File: Tests/Fixtures/TestProject/project.yml
Comment (yonaskolb) [inline Tests/Fixtures/TestProject/project.yml]: Can we add more properties to this to get a fuller integration test

File: Tests/XcodeGenKitTests/BreakpointGeneratorTests.swift
Comment (yonaskolb) [inline Tests/XcodeGenKitTests/BreakpointGeneratorTests.swift]: The current codebase has since moved to using `unwrap`
```swift
let xcbreakpoint = try unwrap(xcodeProject.sharedData?.breakpoints?.breakpoints.first)
```

File: CHANGELOG.md
Comment (yonaskolb) [inline CHANGELOG.md]:
```suggestion
- Add shared breakpoints support [#693](https://github.com/yonaskolb/XcodeGen/pull/693) @myihsan @alexruperez
```

Comment (myihsan) [reply CHANGELOG.md]: Fixed.
Sorry for this terrible mistake.

**Comment by yonaskolb** [comment]
Thanks for the updates @myihsan! There are still a few unresolved comments. Once those are taken care of we can merge this in

**Comment by myihsan** [comment]
Thanks for your review, @yonaskolb!

I want to confirm with you before going on.
For the same reason to change `ignoreCount` to `Int`, I think it might be better to change File Breakpoint as below.

### File Breakpoint

#### Current

- [ ] **filePath**: **String** - breakpoint file path (only used by file breakpoints)
- [ ] **startingColumn**: **Int** - breakpoint starting column (only used by file breakpoints)
- [ ] **endingColumn**: **Int** - breakpoint ending column (only used by file breakpoints)
- [ ] **startingLine**: **Int** - breakpoint starting line (only used by file breakpoints)
- [ ] **endingLine**: **Int** - breakpoint ending line (only used by file breakpoints)

#### Proposal

- [x] **filePath**: **String** - breakpoint file path (only required by file breakpoints)
- [ ] **line**: **Int** - breakpoint line (only used by file breakpoints)

``` swift
startingColumn = "9223372036854775807" // Default value set by Xcode
endingColumn = "9223372036854775807" // Default value set by Xcode
startingLine = String(line)
endingLine = String(line)
```

What do you think?

File: Tests/XcodeGenKitTests/BreakpointGeneratorTests.swift
Comment (myihsan) [reply Tests/XcodeGenKitTests/BreakpointGeneratorTests.swift]: @yonaskolb 
I will fix this issue at the end to avoid resolving conflicts multiple times.
Is it fine to merge `master` to here?
Or should I rebase `master` to here?

Comment (yonaskolb) [reply Tests/XcodeGenKitTests/BreakpointGeneratorTests.swift]: Merging master instead of rebasing is good, so the changes are visible

**Comment by yonaskolb** [comment]
@myihsan yes I agree, it would be good to only need `line` and `filePath`. Perhaps we can keep the other properties but give them them defaults if they are not provided. Otherwise I'd be ok with just leaving them out of the spec.
Also is the default for `startingColumn` not 0?

**Comment by myihsan** [comment]
@yonaskolb 
About the other properties, as my tests on Xcode 11.3.1, they are reset as below by Xcode after running.

``` swift
startingColumn = "9223372036854775807" // not 0
endingColumn = "9223372036854775807"
endingLine = startingLine
```

So I think it may be better to only have `filePath` and `line` to avoid confusion.
If there is somebody needs those properties and knows how to use them, we can discuss it again in an issue.

I also have done some other updates, if you have time please review them first to help me avoid some mistakes in reimplementation for File Breakpoint and adding some tests.

File: Sources/XcodeGenKit/BreakpointGenerator.swift
Comment (yonaskolb) [inline Sources/XcodeGenKit/BreakpointGenerator.swift]: Doesn't need to change but this can be expressed more simply as:
```swift
breakpoint.scope?.rawValue
```

File: Sources/ProjectSpec/Breakpoint.swift
Comment (yonaskolb) [inline Sources/ProjectSpec/Breakpoint.swift]: This is a change in behaviour. It used to be optional, but now has `.throw` as default. If that truly is the default, we should make it a non optional. Maybe keeping default as nil is better though?

File: Tests/Fixtures/TestProject/Project.xcodeproj/xcshareddata/xcdebugger/Breakpoints_v2.xcbkptlist
Comment (yonaskolb) [inline Tests/Fixtures/TestProject/Project.xcodeproj/xcshareddata/xcdebugger/Breakpoints_v2.xcbkptlist]: These properties now have a default. Is that correct? What does Xcode do if you create a breakpoint?

File: Sources/ProjectSpec/Breakpoint.swift
Comment (yonaskolb) [inline Sources/ProjectSpec/Breakpoint.swift]: Where did these values come from?

**Comment by yonaskolb** [comment]
@myihsan yeah I'm ok with leaving them out

File: Sources/ProjectSpec/Breakpoint.swift
Comment (myihsan) [reply Sources/ProjectSpec/Breakpoint.swift]: @yonaskolb 
> If that truly is the default, we should make it a non optional.

Could you tell me the reason?

#### Here is my thought:
Basuse `scrop` and `stopOnStyle` are only required when `type == .exception`, and has a default value, I should do something like below for `scrop` and `stopOnStyle` when `type == .exception`.

``` swift
enabled = jsonDictionary.json(atKeyPath: "enabled") ?? true
```

File: Tests/Fixtures/TestProject/Project.xcodeproj/xcshareddata/xcdebugger/Breakpoints_v2.xcbkptlist
Comment (myihsan) [reply Tests/Fixtures/TestProject/Project.xcodeproj/xcshareddata/xcdebugger/Breakpoints_v2.xcbkptlist]: @yonaskolb 
If I create an Exception Breakpoint,  Xcode set `scope` to `"1"` and `stopOnStyle` to `"0"`.

File: Sources/ProjectSpec/Breakpoint.swift
Comment (myihsan) [reply Sources/ProjectSpec/Breakpoint.swift]: @yonaskolb 
The editing pop-up of an Exception Breakpoint has two single choice properties as below.
- Exception
  - All
  - Objective-C
  - C++
- Break
  - On Throw
  - On Catch

**Comment by myihsan** [comment]
@yonaskolb 
I have done some updates. Please review them.

I also find some properties should be removed. Please tell me what do you think.
- Breakpoint
  - timestamp: I don't know what is this property for and how to set it from Xcode
  - breakpointStackSelectionBehavior: Set to `"1"` by Xcode on some breakpoint types, but I don't know what is this property for and how to change it from Xcode
- Action
  - type
    - OpenGLError: An OpenGLError action will auto appear in an OpenGLError breakpoint

File: Docs/ProjectSpec.md
Comment (yonaskolb) [inline Docs/ProjectSpec.md]: Could we remove the `Breakpoint` suffix from all these, and just added it when encoding/decoding

File: Sources/ProjectSpec/Breakpoint.swift
Comment (yonaskolb) [inline Sources/ProjectSpec/Breakpoint.swift]: To make the API nicer, we could possibly also change the breakpoint type to a custom enum with associated types. This will make creating breakpoints in Swift much friendlier. Not required here in this PR but would be nice. We'd have to be careful to not make incorrect assumptions about which properties belong with which types. What do you think? 

eg:
```swift
type: .file(path: "File.swift", line: 58)
```

File: CHANGELOG.md
Comment (yonaskolb) [inline CHANGELOG.md]: This will have to be moved up into the next version

File: Docs/ProjectSpec.md
Comment (yonaskolb) [reply Docs/ProjectSpec.md]: It looks like it's an easy add in XCProjExtensions as that there's already an init that removes the `Xcode.Breakpoint.` prefix

File: Docs/ProjectSpec.md
Comment (yonaskolb) [inline Docs/ProjectSpec.md]: Can we lowercase these

Comment (myihsan) [reply Docs/ProjectSpec.md]: We can if we restore them before setting them to `XcodeProj `.

File: Sources/ProjectSpec/Breakpoint.swift
Comment (myihsan) [reply Sources/ProjectSpec/Breakpoint.swift]: I thought about it. 
However, I didn't know if I should do it.

I would be glad to do it in this PR.
But it might take some time since I am a little busy with my work.
Is that alright?

File: Docs/ProjectSpec.md
Comment (myihsan) [reply Docs/ProjectSpec.md]: You are right, I will remove the `Breakpoint` suffix.

**Comment by SebastianThiebaud** [comment]
@myihsan I'm more than happy to help / jump on this to speed things up. I'm really excited about this feature.

File: Docs/ProjectSpec.md
Comment (myihsan) [reply Docs/ProjectSpec.md]: Can I ask the reason?
There are some possible values that are not lowercase, so I am confused.
Are we going to make possible values lowercase from now on?

**Comment by myihsan** [comment]
@SebastianThiebaud 
Please try this feature from my branch and tell me if there is anything that should be improved.

**Comment by SebastianThiebaud** [comment]
> @SebastianThiebaud
> Please try this feature from my branch and tell me if there is anything that should be improved.

I built from your repo and branch with `make install`, when running xcodegen with this configuration:
```
    breakpoints:
      - type: Symbolic
        symbol: UIViewAlertForUnsatisfiableConstraints
        actions:
          - type: DebuggerCommand
            command: "expr -l objc++ -O -- [[UIWindow keyWindow] _autolayoutTrace]"
```
It doesn't seem to create a breakpoint. Going to dig deeper, but just to make sure, this configuration is correct, right?

**Comment by myihsan** [comment]
@SebastianThiebaud 
I think it's abnormal to have indent before `breakpoints` since it's a `Project` property.
Did you set it to a target or something? If so, you should remove the indent to make it a `Project` property.

**Comment by SebastianThiebaud** [comment]
@myihsan Good point... It works properly. 
Screen Shot 2020-02-08 at 6 54 57 PM

**Comment by SebastianThiebaud** [comment]
@yonaskolb is there anything here that needs more work / review? sorry to be pushy, I would just hate seeing this PR going to waste... I can't wait for it to be merged 😍

**Comment by myihsan** [comment]
Sorry for everyone who is waiting for this feature. I had forgotten this PR completely🙇‍♂️
Let me check if this feature still does not exist. If so, I will try to rebase to solve the conflict this weekend.

**Comment by myihsan** [comment]
The rebase seems fine.
I noticed that it would be better to support column breakpoints.
I will try it before requesting a review again.

**Comment by myihsan** [comment]
@yonaskolb 
I solved the conflicts by rebasing to the master branch. Please review the changes when you have time.

## About New Features

###  Column Breakpoints

I tried to support it in this pull request [here](https://github.com/yonaskolb/XcodeGen/pull/693/commits/69129c23680d593a95957f6d6ddabaa6f5fc0b0e).
If there is something to discuss, I can drop it and open another pull request.

### Other New Features

To support other new features like Runtime Issue Breakpoint, we have to contribute to https://github.com/tuist/XcodeProj first.
So I am not going to implement them in this pull request.

File: Docs/ProjectSpec.md
Comment (yonaskolb) [reply Docs/ProjectSpec.md:228]: You're right. These don't need to follow swift lower camel casing, and other enums here use uppercase.

File: Docs/ProjectSpec.md
Comment (yonaskolb) [inline Docs/ProjectSpec.md:184]: Let's remove this, as it's not used

File: Sources/ProjectSpec/Breakpoint.swift
Comment (yonaskolb) [inline Sources/ProjectSpec/Breakpoint.swift:98]: Let's remove this as not used

**Review by yonaskolb — COMMENTED** [review]
Great work @myihsan! This was a large undertaking so well done, and thank you for coming back to it 🙏 
Looks good to me, just noticed that references to `timestamp` were still there. After that we can get it merged 😄

**Comment by myihsan** [comment]
@yonaskolb 
Thank you! I also noticed that references to `breakpointStackSelectionBehavior` were still there.
So I fix them in 1c018bae0d507940279a2e931a754a58f83759cf by rebasing for future revert convenience.

**Review by yonaskolb — APPROVED** [review]
Merging 🎉

**Comment by teameh** [comment]
This addition would be really useful, anything I can do to help?

**Comment by yonaskolb** [comment]
@teameh if you could resolve the conflicts here that would be great. Not sure what the hold up was or why this didn't get merged

**Comment by teameh** [comment]
Done. @myihsan see the merge commit [here](https://github.com/teameh/XcodeGen/commit/d27d0343b41c744d1a00bb4a856d499e2c9b8748). Could you reset to that commit? 

<details><summary>For example with:</summary>
<p>

```
git remote add teameh https://github.com/teameh/XcodeGen.git
git fetch teameh
git reset teameh/feature/breakpoint-support --hard
git push
```

</p>
</details>

**Comment by myihsan** [comment]
Thanks, @teameh !
I would like to reset to your commit, but there is a conflict (didn't count as a Git conflict) in `CHANGELOG.md`.
So I tried to solve the conflicts and pushed, @yonaskolb @teameh please review the changes .

**Comment by teameh** [comment]
Looks good!

**Review by yonaskolb — APPROVED** [review]

**Comment by yonaskolb** [comment]
Thanks @myihsan! We got there in the end 😄

**Comment by alexruperez** [comment]
Thank you @myihsan, @yonaskolb!! 🫶

**Pull request merged**
